src/lib/kimyi.rs

```diff
@@ -0,0 +1,737 @@
+                la_idx: n.la_idx,
+                t: 1,
+                repairs: n_repairs,
+                cf: n.cf + cost,
+                cg: 0};
+            nbrs.insert(nn);
+        }
+    }
+}
+
+fn r3d<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
+      (parser: &Parser<TokId>, n: &PathFNode, nbrs: &mut HashSet<PathFNode>)
+{
+    if n.t != 1 || n.la_idx == parser.lexemes.len() {
+        return;
+    }
+
+    let (_, la_term) = parser.next_lexeme(None, n.la_idx);
+    let nn = PathFNode{pstack: n.pstack.clone(),
+                       la_idx: n.la_idx + 1,
+                       t: 1,
+                       repairs: n.repairs.child(ParseRepair::Delete),
+                       cf: n.cf + parser.dc(la_term),
+                       cg: 0};
+    nbrs.insert(nn);
+}
+
+fn r3s_n<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
+      (parser: &Parser<TokId>, n: &PathFNode, nbrs: &mut HashSet<PathFNode>)
+{
+    let (new_la_idx, n_pstack) = parser.lr_cactus(None,
+                                                  n.la_idx,
+                                                  n.la_idx + 1,
+                                                  n.pstack.clone(),
+                                                  &mut None);
+    if new_la_idx == n.la_idx + 1 {
+        let nn = PathFNode{
+            pstack: n_pstack,
+            la_idx: new_la_idx,
+            t: 1,
+            repairs: n.repairs.child(ParseRepair::Shift),
+            cf: n.cf,
+            cg: 0};
+        nbrs.insert(nn);
+    }
+}
+
+/// Convert the output from `astar` into something more usable.
+fn collect_repairs(mut rprs: Vec<PathFNode>) -> Vec<ParseRepair>
+{
+    let mut y = rprs.pop()
+                    .unwrap()
+                    .repairs
+                    .vals()
+                    .cloned()
+                    .collect::<Vec<ParseRepair>>();
+    y.reverse();
+    y
+}
+
+/// Take a vector of parse repairs and return a simplified version. Note that the caller must make
+/// no assumptions about the size or contents of the output: this function might delete, expand, or
+/// do other things to repairs.
+fn simplify_repairs<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
+                   (parser: &Parser<TokId>,
+                    mut rprs: Vec<ParseRepair>)
+                 -> Vec<ParseRepair>
+{
+    // Remove shifts from the end of repairs
+    let sg = parser.grm.sentence_generator(|x| parser.ic(Symbol::Term(x)));
+    while rprs.len() > 0 {
+        if let ParseRepair::Shift = rprs[rprs.len() - 1] {
+            rprs.pop();
+        } else {
+            break;
+        }
+    }
+
+    // Remove all inserts of nonterms which have a minimal sentence cost of 0.
+    let mut j = 0;
+    while j < rprs.len() {
+        if let ParseRepair::InsertNonterm{nonterm_idx} = rprs[j] {
+            let msc = sg.min_sentence_cost(nonterm_idx);
```

Comment:
I'm personally not bothered by this but I'm unsure if this is ok rust syntax, as msc is only used once, so there's no real need to assign this variable.

Reply:
Good point.